File: tps/tp1/questions/q05/sum_digits.c
/*
 * Nome: Lucas Santiago Pereira
 * Matricula: 897498
 * Data: 23/03/2026
*/

// Bibliotecas
#include <stdio.h>
#include <stdlib.h>

// Definicao para null = NULL
#define null NULL

/*
 * Funcao recursiva para retornar soma dos digitos de um numero inteiro
 * @param value valor inteiro
 * @param x variavel para auxiliar na execucao
*/
int sumOfDigits(int value, int x) {

    int k = 0; // Variavel de retorno

    if (x > value) return k; // Caso base
    else {

        k = ((value % (x * 10)) - value % x) / x; // Formula para separar algarismos do numero inteiro
        return k = k + sumOfDigits(value, x * 10); // Motor

    }

}

/*
 * Funcao de encapsulamento para simplificar chamada
 * @param value valor inteiro
*/
int sumOfDigitsEncap(int value) {

    int x = sumOfDigits(value, 1); // Chamada da funcao recursiva com constante 1
    return x; // Retorno

}

/*
 * Funcao para transformar uma cadeia de caracteres em um numero inteiro
 * @param str cadeia de caracteres
 * @param num numero inteiro a ser gerado
*/
int toInt(const char* str, int num) {

    if (*str == '\0') return num; // Caso base

    // str + 1 = proximo valor da cadeia (str[x + 1])
    // *str - '0' = valor da posicao (str[x] - '0')
    return toInt(str + 1, num * 10 + (*str - '0')); // Motor

}

/*
 * Funcao para encapsular
 * @param str string a ser convertida para inteiro
*/
int toIntEncap(const char* str) {

    return toInt(str, 0); // Retornar

}

/*
 * Funcao para comparar dois arranjos de caracteres
 * @param str1 primeiro arranjo a ser comparado
 * @param str2 segundo arranjo a ser comparado
*/
int strCompare(char* str1, char* str2) {

    int OK = 1; // Variavel para controle

    int x = 0; // Variavel para iterar
    while (OK == 1 && str1[x] != '\0' && str2[x] != '\0') {

        if (str1[x] != str2[x]) OK = 0; // Comparar se caracter eh igual
        x = x + 1; // Soma um

    }

    // CORREÇÃO: Garante que os tamanhos sejam iguais para validar
    if (str1[x] != str2[x]) OK = 0;

    return OK; // Retornar

}

/*
 * Funcao principal do programa
 * @param argc contagem de argumentos
 * @param argv argumentos do programa
*/
int main(int argc, char* argv[]) {

    char* resp = malloc(sizeof(char) * 100); // Alocar espaco com tamanho 100

    // CORREÇÃO: Loop infinito controlado para lidar com o Fim de Arquivo (EOF)
    while (1) {
        resp[0] = '\0';
        int lido = scanf(" %99[^\n]", resp); // Ler linha

        if (lido == EOF) break; // Sai do loop se o arquivo terminar

        // CORREÇÃO: Tratamento para ignorar o 'Carriage Return' (\r) do Windows
        int len = 0;
        while(resp[len] != '\0') len++;
        if(len > 0 && resp[len-1] == '\r') resp[len-1] = '\0';

        if (strCompare(resp, "FIM") == 1) break; // Quebra caso encontre a palavra FIM

        int x = toIntEncap(resp); // Tranformar string em inteiro
        printf("%d\n", sumOfDigitsEncap(x)); // Imprimir soma dos digitos
    }

    free(resp); // Boa prática: liberar o ponteiro antes de encerrar
    resp = null;

    return 0;
}

File: tps/tp1/questions/q12/SumDigits.java
/*
 * @author Lucas Santiago Pereira - 897498
 * @date 26/03/2026
*/

import java.io.*;
import java.nio.charset.*;

/*
 * Classe MyIO reduzida com apenas o necessario para o exercicio
*/
class MyIO {

    private static BufferedReader in = new BufferedReader(
        new InputStreamReader(System.in, Charset.forName("ISO-8859-1"))
    );
    private static String charset = "ISO-8859-1";

    public static void setCharset(String charset_){
        charset = charset_;
        in = new BufferedReader(new InputStreamReader(System.in, Charset.forName(charset)));
    }

    public static void print(String x){
        try {
            PrintStream out = new PrintStream(System.out, true, charset);
            out.print(x);
        } catch(UnsupportedEncodingException e){
            System.out.println("Erro: charset invalido");
        }
    }

    public static void println(String x){
        try {
            PrintStream out = new PrintStream(System.out, true, charset);
            out.println(x);
        } catch(UnsupportedEncodingException e){
            System.out.println("Erro: charset invalido");
        }
    }

    public static String readString(){
        String s = "";
        char tmp;
        try{
            do{
                int val = in.read();
                if (val == -1) break; // CORREÇÃO: Tratamento de Fim de Arquivo (EOF)
                tmp = (char) val;
                
                if(tmp != '\n' && tmp != ' ' && tmp != 13){
                    s += tmp;
                }
            }while(tmp != '\n' && tmp != ' ');
        }catch(IOException ioe){
            System.out.println("lerString: " + ioe.getMessage());
        }
        return s;
    }

    public static String readLine(){
        String s = "";
        char tmp;
        try{
            do{
                int val = in.read();
                if (val == -1) break; // CORREÇÃO: Tratamento de Fim de Arquivo (EOF)
                tmp = (char) val;
                
                if(tmp != '\n' && tmp != 13){
                    s += tmp;
                }
            }while(tmp != '\n');
        }catch(IOException ioe){
            System.out.println("lerLine: " + ioe.getMessage());
        }
        return s;
    }

    public static String readString(String prompt){
        print(prompt);
        return readString();
    }

    public static String readLine(String prompt){
        print(prompt);
        return readLine();
    }
}

public class SumDigits {

    /*
     * Funcao para identificar se duas strings sao iguais
     * @param msg1 primeira mensagem a ser comparada
     * @param msg2 segunda mensagem a ser comparada
    */
    public static boolean isEquals(String msg1, String msg2) {

        boolean OK = true; // Variavel de controle

        // Verificar se ambas as strings possuem o mesmo tamanho
        if (msg1.length() == msg2.length()) {

            // Loop para comparar strings
            for (int x = 0; x < msg1.length() && OK; x = x + 1) {

                if (msg1.charAt(x) != msg2.charAt(x)) OK = false; // Teste dos caracteres

            }

        } else OK = false; 

        return OK; 

    }

    /*
     * Funcao recursiva para retornar soma dos digitos de um numero inteiro
     * @param value valor inteiro
     * @param x variavel para auxiliar na execucao
    */
    public static int sumOfDigits(int value, int x) {

        int k = 0; // Variavel de retorno

        if (x > value) return k; // Caso base
        else {

            // Formula para separar algarismos do numero inteiro
            k = ((value % (x * 10)) - value % x) / x; 
            return k + sumOfDigits(value, x * 10); // Motor recursivo

        }

    }

    /*
     * Funcao de encapsulamento para simplificar chamada
     * @param value valor inteiro
    */
    public static int sumOfDigitsEncap(int value) {

        // Chamada da funcao recursiva iniciando o multiplicador em 1
        return sumOfDigits(value, 1); 

    }

    // Metodo principal do programa
    public static void main(String[] args) {

        // String de entrada
        String str = MyIO.readLine();

        // Loop para entrada e processamento
        // O loop encerra se ler "FIM" ou se readLine retornar vazio (EOF)
        while (str != null && !str.equals("") && !(isEquals(str, "FIM"))) { 
         
            // Converter string para inteiro e calcular soma de forma recursiva
            int num = Integer.parseInt(str);
            MyIO.println("" + sumOfDigitsEncap(num));

            str = MyIO.readLine(); // Solicita novamente a entrada

        }

    }

}
